### ISSUE TITLE
Support symlinks for `comports()`

### ISSUE THREAD

**Issue by altendky** [pyserial/pyserial#153]
Since I use multiple serial adapters I have setup `udev` to create symlinks with useful (for me at least) names to keep track of which is which.  For completeness, I am identifying them by the driver and the USB 'path'.

```
ACTION=="add", DRIVERS=="pl2303", DEVPATH=="*/2-3.3.1/*", SYMLINK+="ttymod", MODE="0666"
ACTION=="add", DRIVERS=="pl2303", DEVPATH=="*/2-3.3.2/*", SYMLINK+="ttyhmi", MODE="0666"
```

I would expect a user (such as myself) to want:
* Have a symlink identified device 'work' like a regular one
* Be able to tell if a given identified device is a symlink and get the real path
* Include symlinks to identified devices even if they don't match the globs
 * `/dev/my_serial_adapter -> /dev/ttyUSB1` should be included as a symlink
* Choice to list:
 * set(regular paths)
 * set(symlink paths)
 * set(symlink paths) + (set(regular paths) - set(symlink paths))

That last one would show every device but list symlinks _instead of_ regular paths when available.

I may code this up myself but I wanted to raise the issue for discussion as well.

**Comment by neilwhelchel** [comment]
This is could be problematic, or at least an issue to think about. Linux uses specific device names for serial ports (among many others), ttySxx is for 8250 (16550) like ports, ttymxcxx is for i.MX serial ports, ttyUSBXX is for USB serial ports that emulate the 8250. The reason for this is because differing hardware has differing capabilities, and the name of the port is intended to tell us which driver is in use, and therefore something about the port. While it is possible to trace the link backwards to the actual devnode and look at that to get some information, this is much more of a mess to pass back to the python programmer that is using the serial port. For example, it is likely not reasonable for PySerial to keep track of differing serial drivers, PySerial should be generic, and it really should be up to the programmer that is using PySerial to determine if the serial driver really has the expected capabilities. 
Here is an example:
In a perfect world, if I were to link /dev/ttymxc0 -> /dev/ttyS0, then use PySerial to open that port by passing in '/dev/ttyS0', then PySerial (should) tell us (by examining the devnode that the link points to) that this port is really not 8250 like and tell us that this port for example does not support mark or space parity like a 8250 port does. But we are likely not in a perfect world, and in this case it probably makes more sense to pass this responsibility to the programmer that is using PySerial. If symlinks are not supported, this forces the "port" to contain some information about the driver, and the programmer that is using PySerial can account for this...
Just my two cents worth. ;)

**Comment by zsquareplusc** [comment]
I guess it would mean looking at all the links, at least under /dev, reading each and look if it points to a serial port device (recursively in case it's a link to a link etc).

If you have specialized apps that expect these two devices, it's also possible to access these directly instead of listing all ports in the first place.

Recently, I usually just filter for USB VID:PID or description.

**Comment by altendky** [comment]
@neilwhelchel, if I read it correctly your basic concern is that the symlink may make it harder to know what the actual device is?  It seems easy enough to follow symlinks and then use the final file name just as if it had been found directly.

@zsquareplusc, i got pulled off this task and I can't find my code at the moment but I sure thought I had written up something to search for it.  I know I could just hard code this but I really hate doing that and figured that other people might like this feature as well.

**Comment by neilwhelchel** [comment]
@altendky (Please keep in mind I am not a PySerial developer, I am a fly on the wall, so I might not be speaking for the intentions of the developers or the project.) Yes, it is quite easy to follow the symlink to determine this. My point is that PySerial should not do this. It is my opinion that the application that is using PySerial should follow the symlinks and work out the device that it is going to use. Just as it is the job of the application that is using PySerial to set the comm parameters, it is also the job of that application to know if the parameters apply to a certain device. The line has to be somewhere, maybe there needs to be an add-on to PySerial that works out these things and contains a database of device capabilities. (This is a big project to keep up to date as there are new devices popping up every day or so.) The big problem here is that the actual Linux serial device API (termios) is BADLY broken, and it literally does not tell you when the hardware is not capable of accepting what settings you want. In most cases, setting unsupported parameters silently fails and leaves the user with patches of missing hair.

**Comment by altendky** [comment]
@neilwhelchel Yeah, sounds like we'll disagree on this one.  pyserial has an entire directory dedicated to making it so application developers don't have to know how to list all available serial ports. Allowing for symlinks within some constraint (within `/dev`?) isn't much to add on top of that.  Automatic identification of serial parameters would likely be in a whole different ballpark so I don't really see an argument there.

There is no additional confusion about properties and characteristics of the device as `os.path.realpath()` will give you the exact same thing that `glob.glob()` does now, the path of the actual device (such as `/dev/ttyUSB0`).  The extra information provided is what is likely a user selected name (in my case `/dev/ttymod`).  To get this there is the minimal extra difficulty of parsing 'all' symlinks.  Maybe just `/dev/tty*` in Linux? Maybe all files in `/dev` and then identify the ones that are symlinks to the devices already found with `glob.glob()`?  Regardless, not a major task.  Again, hopefully I'll write up a patch myself when I get back to this project at work.

**Comment by neilwhelchel** [comment]
@altendky It seems that you might have misunderstood what I was suggesting. I was suggesting that a module be added to PySerial that is a higher level API than PySerial that does what you were suggesting. (That is what I meant by on top of PySerial.) This API could end up looking similar in a few places but it could add another layer of things like device capabilities, actual serial protocols like 9 bit A/D, YASP, MODBUS, etc. The issue is that many of these protocols require specific timing and specific control of the serial hardware. This requires actual knowledge of the underlying hardware in some cases to even make it work correctly. PySerial is realistically about the lowest level API that you can go to the serial port and still be pythonic, so it makes sense to set another layer on top of it that gives you more tools and leave the lower level API as stripped down as possible, yet get full control over the hardware. (Again, I am not trying to speak for the PySerial developers, I am just talking from my 35 years of serial protocol experience and a moderate amount of Linux kernel serial driver programming.)

**Comment by altendky** [comment]
@neilwhelchel I believe I did misunderstand so thanks for the clarification.  That said, I would not think that implementing protocols within PySerial would make the most sense.

In case it helps to understand why I think this feature would be a reasonable addition within the existing code, you can see the extent of the core of my suggestion in https://gist.github.com/44b8cef79c419e50bade73a144c4386e (26 lines added).  I think this implements everything except my last request for configurability in what is reported.  It would probably be a good idea to maintain existing functionality by default and various other considerations I haven't put in but those are likely minor tweaks.

I 'tested' (very loosely using this term) as follows.  Note that the existing `.device` member variable retains the real path and so any detection based on that path done in the rest of the library should still work.  `.link` is provided so the application can present that instead of it prefers.  There likely could be some function to get the 'name' or 'path' or... or perhaps that already exists and should just be tweaked.

```
(venv) altendky@tp:~/ps$ python -c 'from serial.tools.list_ports import comports; print("\n".join([" -> ".join([d for d in [p.link, p.device] if d is not None]) for p in comports()]))'
/dev/ttyUSB0
/dev/ttyUSB2
/dev/ttyUSB1
/dev/ttymod -> /dev/ttyUSB1
/dev/ttyhmi -> /dev/ttyUSB2
```

**Comment by kurtmckee** [comment]
@zsquareplusc @projectgus Please close this issue. It was fixed by 1c19c42 but didn't get closed at that time.
